src/main/java/org/wildfly/security/http/impl/BasicAuthenticationMechanism.java

```diff
@@ -0,0 +1,169 @@
+import static org.wildfly.security.http.HttpConstants.AUTHORIZATION;
+import static org.wildfly.security.http.HttpConstants.BASIC;
+import static org.wildfly.security.http.HttpConstants.CHARSET;
+import static org.wildfly.security.http.HttpConstants.REALM;
+import static org.wildfly.security.http.HttpConstants.UNAUTHORIZED;
+import static org.wildfly.security.http.HttpConstants.WWW_AUTHENTICATE;
+import static org.wildfly.security.util._private.Arrays2.indexOf;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.util.List;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+
+import org.wildfly.security.auth.callback.AuthenticationCompleteCallback;
+import org.wildfly.security.auth.callback.PasswordVerifyCallback;
+import org.wildfly.security.auth.callback.SecurityIdentityCallback;
+import org.wildfly.security.http.HttpAuthenticationException;
+import org.wildfly.security.http.HttpServerAuthenticationMechanism;
+import org.wildfly.security.http.HttpServerExchange;
+import org.wildfly.security.util.ByteIterator;
+
+/**
+ * Implementation of the HTTP BASIC authentication mechanism
+ *
+ * @author <a href="mailto:darran.lofthouse@jboss.com">Darran Lofthouse</a>
+ */
+class BasicAuthenticationMechanism implements HttpServerAuthenticationMechanism {
+
+    // TODO - Undertow also has a silent mode for HTTP authentication.
+
+    private static final String BASIC_PREFIX = BASIC + " ";
+    private static final int PREFIX_LENGTH = BASIC_PREFIX.length();
+
+    private final CallbackHandler callbackHandler;
+    private final String challengeValue;
+
+    BasicAuthenticationMechanism(final CallbackHandler callbackHandler, final String realm, final boolean includeCharset) {
+        checkNotNullParam("callbackHandler", callbackHandler);
+        checkNotNullParam("realm", realm);
+
+        this.callbackHandler = callbackHandler;
+
+        StringBuilder sb = new StringBuilder(BASIC_PREFIX);
+        sb.append(REALM).append("=\"").append(realm).append("\"");
+        if (includeCharset) {
+            sb.append(", ").append(CHARSET).append("=\"UTF-8\"");
+        }
+        challengeValue = sb.toString();
+
+    }
+
+    /**
+     * @see org.wildfly.security.http.HttpServerAuthenticationMechanism#getMechanismName()
+     */
+    @Override
+    public String getMechanismName() {
+        return BASIC;
+    }
+
+    /**
+     * @throws HttpAuthenticationException
+     * @see org.wildfly.security.http.HttpServerAuthenticationMechanism#evaluateRequest(org.wildfly.security.http.HttpRequest)
+     */
+    @Override
+    public boolean evaluateRequest(HttpServerExchange exchange) throws HttpAuthenticationException {
+        List<String> authorizationValues = exchange.getRequestHeaderValues(AUTHORIZATION);
```

Comment:
As you say it is not clear, this check and loop is to cope with the possibility that there could be more than one - not an expectation that there is more than one.

IMO for a situation like this we either need to cope with the possibility that this returns multiple headers or report an error if we encounter mutliple headers.  Keeping those two in mind my judgement is that reporting an error is the scenario most likely to result in a bug report from a user if they do encounter it.

For the single header cases this loop works fine and in a situation where multiple headers are present this loop will also handle that.